### PR TITLE
[bitnami/mastodon] feat: :tada: Add chart

### PR DESCRIPTION
**Pull request by javsalgar**
Signed-off-by: Javier Salmeron Garcia <jsalmeron@vmware.com>

### Description of the change

This PR provides a helm chart for Mastodon. Features:

 - Bitnami standards
 - Ingress support via the Apache chart
 - Flexibility of the configuration
 - Scalability by default thanks to the MinIO chart 
 - Init jobs that performs the initialization logic separate from the deployments
 - Admin user creation

Has the following requirements:

 - PostgreSQL, Redis, Elasticsearch, Apache and MinIO as subcharts
 - Requires a domain to be previously set

 
### Benefits

New popular solution to our catalog

### Possible drawbacks

n/a

### Applicable issues

  - fixes #

### Checklist

- [ ] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/). This is *not necessary* when the changes only affect README.md files.
- [ ] Variables are documented in the values.yaml and added to the `README.md` using [readme-generator-for-helm](https://github.com/bitnami-labs/readme-generator-for-helm)
- [ ] Title of the pull request follows this pattern [bitnami/<name_of_the_chart>] Descriptive title
- [ ] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/main/CONTRIBUTING.md#sign-your-work)

### CODE REVIEW

File: bitnami/mastodon/templates/init-job/init-job.yaml [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+                  name: {{ include "mastodon.elasticsearch.secretName" . }}
+                  key: {{ include "mastodon.elasticsearch.passwordKey" . | quote }}
+            {{- end }}
+            {{- if .Values.initJob.extraEnvVars }}
+            {{- include "common.tplvalues.render" (dict "value" .Values.initJob.extraEnvVars "context" $) | nindent 12 }}
+            {{- end }}
+          envFrom:
+            - configMapRef:
+                name: {{ include "mastodon.defaultConfigmapName" . }}
+            - configMapRef:
```
Comment (fmulero) [inline bitnami/mastodon/templates/init-job/init-job.yaml:82]: I think this the reason why this configmap is created empty...

**Review by fmulero — COMMENTED** [review]
Great!!!! Thanks a lot @javsalgar

I left some comments, Could you take them a glance?

File: bitnami/mastodon/templates/init-job/init-job.yaml (hunk 1 above)
Comment (javsalgar) [reply bitnami/mastodon/templates/init-job/init-job.yaml:82]: Fixed

**Review by fmulero — DISMISSED** [review]
LGTM

File: .vib/mastodon/vib-verify.json [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+          "url": "{SHA_ARCHIVE}",
+          "path": "/bitnami/mastodon"
+        },
+        "runtime_parameters": "YWRtaW5FbWFpbDogInZpYi11c2VyQGV4YW1wbGUuY29tIgphZG1pblBhc3N3b3JkOiAiYml0bmFtaSExMjM0IgoKd2ViRG9tYWluOiBiaXRuYW1pLW1hc3RvZG9uLm15CgpzZXJ2aWNlQWNjb3VudDoKICBhdXRvbW91bnRTZXJ2aWNlQWNjb3VudFRva2VuOiB0cnVlCgp3ZWI6CiAgY29udGFpbmVyUG9ydHM6CiAgICBodHRwOiA4MTIzCiAgY29udGFpbmVyU2VjdXJpdHlDb250ZXh0OgogICAgZW5hYmxlZDogdHJ1ZQogICAgcnVuQXNVc2VyOiAxMDAyCiAgcG9kU2VjdXJpdHlDb250ZXh0OgogICAgZW5hYmxlZDogdHJ1ZQogICAgZnNHcm91cDogMTAwMgogIHNlcnZpY2U6CiAgICBwb3J0czoKICAgICAgaHR0cDogODUwMAoKc3RyZWFtaW5nOgogIHNlcnZpY2U6CiAgICBwb3J0czoKICAgICAgaHR0cDogODYwMAoKYXBhY2hlOgogIGVuYWJsZWQ6IHRydWUKICBzZXJ2aWNlOgogICAgdHlwZTogTG9hZEJhbGFuY2VyCiAgICBwb3J0czoKICAgICAgaHR0cDogODAKCnBvc3RncmVzcWw6CiAgZW5hYmxlZDogdHJ1ZQogIHByaW1hcnk6CiAgICBzZXJ2aWNlOgogICAgICBwb3J0czoKICAgICAgICBwb3N0Z3Jlc3FsOiA5MzkzCgptaW5pbzoKICBlbmFibGVkOiB0cnVlCiAgc2VydmljZToKICAgIHBvcnRzOgogICAgICBhcGk6IDgwMTEKCnJlZGlzOgogIGVuYWJsZWQ6IHRydWUKICBtYXN0ZXI6CiAgICBzZXJ2aWNlOgogICAgICBwb3J0czoKICAgICAgICByZWRpczogNjgwMAoKZWxhc3RpY3NlYXJjaDoKICBlbmFibGVkOiB0cnVlCiAgc2VydmljZToKICAgIHBvcnRzOgogICAgICByZXN0QVBJOiA3Nzc3Cg==",
```
Comment (FraPazGal) [inline .vib/mastodon/vib-verify.json:25]: I would add some more parameters here to expand our goss tests:

- `serviceAccount.create=true`: added always just to be sure the SA mounted is created by the chart.
- `web.containerSecurityContext.capabilities.drop=["ALL"]` and its goss test `check-no-capabilities` used in several asset (metallb, argo-cd, etc)
- `web.podSecurityContext.seccompProfile.type="RuntimeDefault"` and its test.
- `persistence.mountPath=/bitnami/mastodon` and its file test in goss similar to the ones used for Cassandra or Solr.

File: .vib/mastodon/goss/goss.yaml [hunk 3]
```diff
@@ -0,0 +1,34 @@
+addr:
```
Comment (FraPazGal) [inline .vib/mastodon/goss/goss.yaml:1]: I think we can remove the `service.ports` addr checks done here for the subcharts. They are only testing that the subcharts can use different ports for their services, and that is something that should be tested in each chart's test suite.

A more interesting test would be to check that mastodon's config is using those non-default ports to connect to the different services (via mastodon's config file or similar). Similar to this, we could also modify the default auth params of each subchart and check the asset is using them correctly.

**Review by FraPazGal — COMMENTED** [review]
Hi @javsalgar! I have added a few comments regarding the testing suite for Mastodon, please check it when possible.

File: .vib/mastodon/cypress/cypress/integration/mastodon_spec.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+      force: true,
+    });
+    cy.get('[type="submit"]').should('be.enabled').click();
+    cy.get('.status-public').contains(`${tweet.text} ${random}`);
+  });
+  // The image will get renamed so we need to use a regex containing the media_attachments section
+  // which will be the only element containing it
+  cy.get('.media-gallery').first().click();
+  cy.get('[src*="media_attachments"]');
+});
```
[the remark this answers is not in the document]
Comment (javsalgar) [reply .vib/mastodon/cypress/cypress/integration/mastodon_spec.js:21]: The thing is that the Elasticsearch ingestion takes some minutes. You publish a tweet and it is not directly accessible in the search engine. Therefore we would need to add a wait and then perform the search.

File: .vib/mastodon/goss/goss.yaml (hunk 3 above)
Comment (javsalgar) [reply .vib/mastodon/goss/goss.yaml:1]: Mastodon does not have a config file, everything is done via environment variables.

File: .vib/mastodon/vib-verify.json (hunk 2 above)
Comment (javsalgar) [reply .vib/mastodon/vib-verify.json:25]: Mastodon by default does not use persistence. This is only necessary when MinIO is not enabled.

File: .vib/mastodon/goss/goss.yaml (hunk 3 above)
Comment (FraPazGal) [reply .vib/mastodon/goss/goss.yaml:1]: In that case, we can check the port-related envs use correct values with the `env` command. That and the check to the `/health` endpoint should be enough.

File: .vib/mastodon/cypress/cypress/integration/mastodon_spec.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (FraPazGal) [reply .vib/mastodon/cypress/cypress/integration/mastodon_spec.js:21]: In that case, we would have to make do with Mastodon's `/health` endpoint used in Goss.

**Review by FraPazGal — APPROVED** [review]
LGTM, thanks for the contribution @javsalgar!

**Pull request merged**